src/coreclr/src/tools/Common/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs

```diff
@@ -169,38 +185,83 @@ public bool TryGetValue(TKey key, out TValue value)
             int hashCode = GetKeyHashCode(key);
             int tableIndex = HashInt1(hashCode) & mask;
 
-            if (hashTableLocal[tableIndex] == null)
+            TValue examineEntry = hashTableLocal[tableIndex];
+            if ((examineEntry == null) || (examineEntry == _entryInProcessOfWritingSentinel))
             {
+                sentinel = Volatile.Read(ref _entryInProcessOfWritingSentinel);
+                if (sentinel != null && CompareKeyToValue(key, sentinel))
+                {
+                    value = sentinel;
+                    return true;
+                }
                 value = null;
                 return false;
             }
 
-            if (CompareKeyToValue(key, hashTableLocal[tableIndex]))
+            if (CompareKeyToValue(key, examineEntry))
             {
-                value = hashTableLocal[tableIndex];
+                value = examineEntry;
                 return true;
             }
 
             int hash2 = HashInt2(hashCode);
             tableIndex = (tableIndex + hash2) & mask;
-
-            while (hashTableLocal[tableIndex] != null)
+            examineEntry = hashTableLocal[tableIndex];
+            while ((examineEntry != null) && (examineEntry != _entryInProcessOfWritingSentinel))
             {
-                if (CompareKeyToValue(key, hashTableLocal[tableIndex]))
+                if (CompareKeyToValue(key, examineEntry))
                 {
-                    value = hashTableLocal[tableIndex];
+                    value = examineEntry;
                     return true;
                 }
                 tableIndex = (tableIndex + hash2) & mask;
+                examineEntry = hashTableLocal[tableIndex];
             }
+
+            sentinel = Volatile.Read(ref _entryInProcessOfWritingSentinel);
+            if (sentinel != null && CompareKeyToValue(key, sentinel))
+            {
+                value = sentinel;
+                return true;
+            }
+
             value = null;
             return false;
         }
 
+        /// <summary>
+        /// Spin and wait for a sentinel to disappear.
+        /// </summary>
+        /// <param name="hashtable"></param>
+        /// <param name="tableIndex"></param>
+        /// <returns>The value that replaced the sentinel, or null</returns>
+        TValue WaitForSentinelInHashtableToDisappear(TValue[] hashtable, int tableIndex)
+        {
+            TValue sentinel = Volatile.Read(ref _entryInProcessOfWritingSentinel);
+            if (sentinel == null)
+                return null;
+
+            TValue value = Volatile.Read(ref hashtable[tableIndex]);
+            while (true)
+            {
+                for (int i = 0; (i < 10000) && value == sentinel; i++)
+                {
+                    value = Volatile.Read(ref hashtable[tableIndex]);
+                }
+                if (value != sentinel)
+                    break;
+
+                Task.Delay(1).Wait();
```

Comment:
`Thread.Sleep(1);`?

Reply:
Or:
```C#
SpinWait sw = default;
while (true)
{
    ...
    sw.SpinOnce();
}
```
?

Reply:
Right. As I understand Yielding/Sleeping options were limited in corert. 
In CoreCLR SpinWait is typically good enough to not roll your own spinning scheme.

Reply:
This code is currently carefully arranged to be compiled against .NET Standard 1.3 which doesn't have any of those apis. I don't currently want to change that.